capsules/extra/src/bmm150.rs

```diff
@@ -0,0 +1,203 @@
+    }
+}
+
+impl<'a, I: i2c::I2CDevice> NineDof<'a> for BMM150<'a, I> {
+    fn set_client(&self, client: &'a dyn NineDofClient) {
+        self.ninedof_client.set(client);
+    }
+
+    fn read_magnetometer(&self) -> Result<(), ErrorCode> {
+        self.start_measurement()
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+enum State {
+    Suspend,
+    Sleep,
+    PowerOn,
+    InitializeReading,
+    ReadMeasurement,
+    Read,
+}
+
+impl<'a, I: I2CDevice> I2CClient for BMM150<'a, I> {
+    fn command_complete(&self, buffer: &'static mut [u8], status: Result<(), i2c::Error>) {
+        if let Err(i2c_err) = status {
+            self.state.set(State::Sleep);
+            self.buffer.replace(buffer);
+            self.ninedof_client
+                .map(|client| client.callback(i2c_err as usize, 0, 0));
+            return;
+        }
+
+        match self.state.get() {
+            State::PowerOn => {
+                buffer[0] = Registers::CTRL2 as u8;
+                buffer[1] = 0x3A_u8;
+
+                if let Err((error, buffer)) = self.i2c.write(buffer, 2) {
+                    self.buffer.replace(buffer);
+                    self.i2c.disable();
+                    self.ninedof_client
+                        .map(|client| client.callback(error as usize, 0, 0));
+                } else {
+                    self.state.set(State::InitializeReading);
+                }
+            }
+            State::InitializeReading => {
+                buffer[0] = Registers::DATAxLsb as u8;
+
+                if let Err((i2c_err, buffer)) = self.i2c.write(buffer, 1) {
+                    self.state.set(State::Sleep);
+                    self.buffer.replace(buffer);
+                    self.ninedof_client
+                        .map(|client| client.callback(i2c_err as usize, 0, 0));
+                } else {
+                    self.state.set(State::ReadMeasurement);
+                }
+            }
+            State::ReadMeasurement => {
+                if let Err((i2c_err, buffer)) = self.i2c.read(buffer, 8) {
+                    self.state.set(State::Sleep);
+                    self.buffer.replace(buffer);
+                    self.ninedof_client
+                        .map(|client| client.callback(i2c_err as usize, 0, 0));
+                } else {
+                    self.state.set(State::Read);
+                }
+            }
+            State::Read => {
+                let x_axis = ((buffer[1] as i16) << 5) | ((buffer[0] as i16) >> 3);
+                let y_axis = ((buffer[3] as i16) << 5) | ((buffer[2] as i16) >> 3);
+                let z_axis = ((buffer[5] as i16) << 7) | ((buffer[4] as i16) >> 1);
+
+                self.buffer.replace(buffer);
+                self.i2c.disable();
+                self.ninedof_client.map(|client| {
+                    client.callback(x_axis as usize, y_axis as usize, z_axis as usize)
+                });
+
+                self.state.set(State::Sleep);
```

Comment:
I think setting the state to sleep needs to happen first. Otherwise, this could happen:

1. client.callback() happens here. State is read.
2. in the callback, the client asks for another measurement by calling read_measurement which calls start_measurement.
3. Start measurement fails because the state is not suspend or sleep.